src/coreclr/jit/priorityqueue.h

```diff
@@ -0,0 +1,110 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// PriorityQueue: A priority queue implemented as a max-heap
+
+template <typename T, typename Compare>
+class PriorityQueue
+{
+private:
+    jitstd::vector<T> data;
+    Compare           comp;
+
+#ifdef DEBUG
+    // Returns true only if each element has a higher priority than its children.
+    bool VerifyMaxHeap() const
+    {
+        for (size_t i = 0; i < data.size(); i++)
+        {
+            const size_t leftChild  = (2 * i) + 1;
+            const size_t rightChild = leftChild + 1;
+
+            if (rightChild < data.size())
+            {
+                if (comp(data[i], data[leftChild]) || comp(data[i], data[rightChild]))
+                {
+                    return false;
+                }
+            }
+            else if (leftChild < data.size())
+            {
+                if (comp(data[i], data[leftChild]))
+                {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+#endif // DEBUG
+
+public:
+    PriorityQueue(const jitstd::allocator<T>& allocator, const Compare& compare)
```

Comment:
Allow for pre-sizing the vector?

Reply:
Currently, `jitstd::vector` does not expose a constructor or `reserve` method to pre-size the underlying array, and I don't see much utility in initializing a priority queue with a default value. Would you like me to extend `jitstd::vector` to support this while we're here?

Reply:
Maybe? For your use case you know up front (or can guess) how many items you'll add... I suppose this can wait until we think we will benefit from it.

Reply:
> For your use case you know up front (or can guess) how many items you'll add

For this, are you suggesting we sum the number of successors out of each block in 3-opt's candidate list to get a loose upper bound for the capacity of `cutPoints`? Since we don't allow inserting duplicate edges, I think this would be a true upper bound, and most methods might have small enough worklists that we can afford this upper bound to be quite liberal. Once 3-opt is in, I can try extending `jitstd::vector`'s API surface and seeing if reserving up-front saves us TP most of the time.